### ISSUE TITLE
Convert sparseNnetDemo to python

### ISSUE THREAD

**Issue by mjsML** [probml/pyprobml#210]
This is figure number 13.21 in the book. its matlab code is [here](https://github.com/probml/pmtk3/blob/master/demos/sparseNnetDemo.m)

**Comment by evcu** [comment]
I like to take on this issue. Is there a python version of the `makeData` functions used in the Matlab script? If not, how can I find the source code?

**Comment by mjsML** [comment]
@evcu thank you for your interest. I believe the function you are referring to is [here](https://github.com/probml/pmtksupport/blob/master/markSchmidt-9march2011/markSchmidt/minFunc/minFunc_examples.m).
IIUC this is part of the default base package for the matlab code or "assumed" to be there.

@murphyk can I ask you to confirm my hypothesis?

**Comment by evcu** [comment]
I found the makeData function and convert it to python. I have few follow-up questions before diving deep.
- Is it ok to use tensorflow library for NNs?
- Regarding the example alone I would like to ask whether following changes are ok: (a) Using Adam optimizer instead of line-search. (b) Using additive l1 instead of projection. (c) Finetuning for few steps after making the network sparse. 
- Do you have preferred way of plotting 1d functions in python? If so maybe I can start from there.
- Do you have any preference/suggestions for plotting sparse NN?
- Do you prefer scripts? or colabs would work?

Best

**Comment by mjsML** [comment]
> I found the makeData function and convert it to python. I have few follow-up questions before diving deep.
> 
> * Is it ok to use tensorflow library for NNs?

There is a slight preference towards [Flax](https://github.com/google/flax), However if TF is your weapon of choice go ahead.

> * Regarding the example alone I would like to ask whether following changes are ok: (a) Using Adam optimizer instead of line-search. (b) Using additive l1 instead of projection. (c) Finetuning for few steps after making the network sparse.

I would defer to @murphyk on this as he is the owner of the figures.

> * Do you have preferred way of plotting 1d functions in python? If so maybe I can start from there.

Not that I'm aware of, @murphyk would also be asked to chime in here , but I believe [this](https://towardsdatascience.com/watching-machine-learning-models-fitting-a-curve-c594fec4bbdb) might be helpful 
> * Do you have any preference/suggestions for plotting sparse NN?
> * Do you prefer scripts? or colabs would work?

Ideally scripts, because this will end up in a larger colab that gives book readers the ability to edit the script and visualize the result e.g [figure 1.5](https://figures.probml.ai/book1/1.5)

> 
> Best

I hope this helps 🙂

**Comment by murphyk** [comment]
The goal is make a very simple neural MLP using JAX  (eg [using stax, as in this code](https://github.com/google/jax/blob/master/examples/mnist_classifier.py), and then use some kind of standard optimizer that can handle L1 penalty (eg projected gradient descent).

**Comment by karalleyna** [comment]
Could I work on this?

**Comment by evcu** [comment]
I was planning on it over the weekend. Happy to pass it after that if I can't finish it over the weekend.

**Comment by karalleyna** [comment]
Since you haven't taken any action for a month, I supposed that this issue was available. Therefore, I pushed my implementation forward yesterday. So, I'm really sorry for that and I don't know what we should do.

**Comment by mjsML** [comment]
@evcu @karalleyna if either of you has a finished implementation (ready for review), please create a pull request.

**Comment by murphyk** [comment]
We can have 2 implementations (with 2 PRs), since there may be more than one way to do this (eg there are multiple approaches to solving L1 regularization - projected gradient, soft relaxation, etc)

**Comment by evcu** [comment]
Sounds good. @karalleyna feel free to take on this. I (or others) can fork your code if necessary to create different examples. If you can't, I can do it after Neurips deadline.

best
